### PR TITLE
Add `get_repetition_penalty_reward`

### PR DESCRIPTION
**Pull request by qgallouedec**
TRL ships only a handful of built-in reward functions (`accuracy_reward`, `reasoning_accuracy_reward`, `think_format_reward`, `get_soft_overlong_punishment`). The RLVR/GRPO literature describes many more rule-based, verifiable rewards that users currently have to reimplement themselves. This PR starts closing that gap with one of the most broadly useful ones.

## What

Adds `get_repetition_penalty_reward`, an n-gram repetition penalty that discourages degenerate, repetitive completions:

```math
R_{\text{repetition}}(y) = \left(1 - \frac{\#\,\text{unique } n\text{-grams}}{\#\,\text{total } n\text{-grams}}\right) \times p
```

A completion with no repeated n-gram gets `0.0`; a fully repetitive one approaches `max_penalty`.

It's meant as an **auxiliary reward** alongside a correctness reward: repetitive output is a common failure mode and a known reward-hacking strategy under length-/format-shaping rewards, so penalizing it stabilizes training.

The n-grams are computed over the completion **token ids** (`completion_ids`), matching the paper, which applies the penalty to repeated tokens. This makes the reward tokenizer-defined and language-agnostic (no whitespace/segmentation assumptions), and consistent with `get_soft_overlong_punishment`, which also consumes `completion_ids`.

```python
from trl.rewards import get_repetition_penalty_reward

repetition_penalty = get_repetition_penalty_reward(ngram_size=3, max_penalty=-1.0)
trainer = GRPOTrainer(..., reward_funcs=[..., repetition_penalty])
```

Reference: Appendix C.2 of [*Demystifying Long Chain-of-Thought Reasoning in LLMs*](https://huggingface.co/papers/2502.03373).

## Notes

- Operates on token ids (`completion_ids`), so it's tokenizer-defined and language-agnostic, with no extra segmentation dependency.
- Implemented as a picklable callable class, matching `get_soft_overlong_punishment` (required by the async GRPO rollout worker).
- Adds a dedicated subsection to `paper_index.md` and an entry to the rewards docs.
- 12 unit tests covering exact penalty values, edge cases (empty / short completions), `max_penalty` scaling, `ngram_size`, batching, positive-`max_penalty` validation, and picklability.

---

> [!NOTE]
> **Low Risk**
> Additive API and documentation with isolated reward logic and tests; no changes to training core paths beyond a new optional reward hook.
> 
> **Overview**
> Adds **`get_repetition_penalty_reward`**, a built-in auxiliary reward that penalizes repeated token n-grams in `completion_ids` (scaled by `max_penalty`, default 3-grams). Intended for use with **`GRPOTrainer`** alongside correctness rewards to reduce repetitive, reward-hacking completions during long CoT RL, per *Demystifying Long Chain-of-Thought Reasoning* (Appendix C.2).
> 
> Implementation mirrors **`get_soft_overlong_punishment`**: a picklable `_RepetitionPenalty` callable class, `**kwargs` ignored for trainer compatibility, and export via **`trl.rewards`**. Docs add a **`paper_index.md`** subsection and **`rewards.md`** autodoc; **12 unit tests** cover scoring, edge cases, validation, and pickling.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 644190ddd9ac0303361028b21004ba1b289da296. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/rewards/other_rewards.py [hunk 1]
```diff
@@ -15,6 +15,71 @@
...
+    >>> repetition_penalty(completion_ids)
+    [0.0, -0.75]
+    ```
+    """
+    if max_penalty > 0:
+        raise ValueError(f"max_penalty {max_penalty} should not be positive")
+    return _RepetitionPenalty(ngram_size, max_penalty)
+
+
+class _RepetitionPenalty:
```
Comment (AmineDiro) [inline trl/rewards/other_rewards.py:62]: I am starting to question if we shouldn't have put the trainer in a separate process ?? 🤔 🤔 🤔  

Not related to this PR, but I can see this being painful for users

Comment (qgallouedec) [reply trl/rewards/other_rewards.py:62]: I'd just for reward function factories that you need this kind of pattern. I think that usually  users just specify the function directly.
Anyway, yes, we can discuss it later

**Review by kashif — APPROVED** [review]
